File: 백준/Silver/1389. 케빈 베이컨의 6단계 법칙/케빈 베이컨의 6단계 법칙.py
import sys
input = sys.stdin.readline

n, m = map(int, input().split())
f = [[0]*(n+1) for _ in range(n+1)]
for _ in range(m):
    a, b = map(int, input().split())
    f[a][b] = 1
    f[b][a] = 1

for k in range(1, n+1):
    for i in range(1, n+1):
        for j in range(1, n+1):
            if i == j: continue
            if f[i][k] != 0 and f[k][j] != 0:
                if f[i][j] == 0:
                    f[i][j] = f[i][k]+f[k][j]
                else: 
                    f[i][j] = min(f[i][j], f[i][k]+f[k][j])

temp = sum(f[1])
result = 1
for n in range(2, n+1):
    if temp > sum(f[n]):
        temp = sum(f[n])
        result = n

print(result)

File: 백준/Silver/1389. 케빈 베이컨의 6단계 법칙/케빈 베이컨의 6단계 법칙.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    private static int N, M, dist[][];
    private static final int INF = 987654321; // 오버플로 조심

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        dist = new int[N][N];
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (i == j) dist[i][j] = 0;
                else dist[i][j] = INF;
            }
        }
        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken()) - 1;
            int b = Integer.parseInt(st.nextToken()) - 1;
            dist[a][b] = 1;
            dist[b][a] = 1;
        }
//        debugPrint();
        for (int k = 0; k < N; k++) {
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
                }
            }
        }
//        debugPrint();
        int result = INF;
        int resultIdx = 0;
        for (int i = 0; i < N; i++) {
            int sum = 0;
            for (int j = 0; j < N; j++) {
                sum += dist[i][j];
            }
            if (result > sum) {
                result = sum;
                resultIdx = i + 1;
            }
        }
        System.out.println(resultIdx);
    }

    private static void debugPrint() {
        for (int i = 0; i < N; i++)
            System.out.println(Arrays.toString(dist[i]));
    }
}